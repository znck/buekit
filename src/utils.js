/**
 * Wrap to array.
 * @param {*} any
 * @returns {Array}
 */
export function toArray(any) {
	return Array.isArray(any) ? any : (
    any === undefined ? [] : [any]
  )
}

const isDot = /\./
export function normalizeProps (
  props,
  resolve = (name, suffix, value) => value,
  transform = (name, value) => value
) {
  const keys = Object.keys(props)
  const base = keys.filter(key => !isDot.test(key))
  const others = keys.filter(key => isDot.test(key))
  const normalized = {}

  base.forEach(
    name => {
      const start = name.length + 1
      normalized[name] = props[name] !== undefined ? transform(name, props[name]) : props[name]
      others.forEach(
        key => {
          if (props[key] !== undefined && key.startsWith(name)) {
            normalized[name] = [
              ...toArray(normalized[name]),
              ...toArray(resolve(name, key.substr(start), props[key]))
            ]
          }
        }
      )
    }
  )

  return normalized
}

/**
 * Merge vue data for functional components.
 * @param {object} local
 * @param {object} parent
 * @returns {object}
 */
export function mergeData(...nodes) {
	const classes = nodes.reduce(
    (result, node) => node && node.class ? result.concat(toArray(node.class)) : result, []
  )

	return Object.assign(...nodes, {class: classes})
}

/**
 * Set value and return target back.
 * @param {object} target
 * @param {string|number} key
 * @param {*} value
 * @returns {objext}
 */
export function fluentSet(target, key, value = true) {
	target[key] = value

	return target
}

/**
 * Convert list of keys to a hash map.
 * @param {Array<string|number>} keys
 * @returns {Object<string, bool>}
 */
export function toMap(keys) {
	return keys.reduce((result, key) => fluentSet(result, key), {})
}

export const responsiveTypes = toMap(['mobile', 'tablet', 'desktop', 'widescreen', 'fullhd'])

/**
 * Verify type is one of responsive types.
 * @param {string} typeName
 * @returns {bool}
 */
export function isResponsiveType(typeName) {
	if (typeName.indexOf('-') > -1) {
		typeName = typeName.split('-').pop()
	}

	return typeName in responsiveTypes
}

/**
 * Valid value with given validator.
 * @param {*} value
 * @param {function} validator
 * @returns {bool}
 */
export function validate(value, validator) {
	return toArray(value).every(validator)
}

export function isNumber(any) {
	return typeof any === 'number'
}

export function isString(any) {
	return typeof any === 'string'
}

export function isBoolean(any) {
	return typeof any === 'boolean'
}

export function isUndefined(any) {
	return typeof any === 'undefined'
}

export function isFunction(any) {
	return typeof any === 'function'
}

export function styleResolver(any) {
	any = any || new Proxy({}, { get: (_, key) => key })

	return key => any[key]
}

export const cssModule = styleResolver

export function isIgnoredVnode(vnode) {
	return !vnode || !vnode.tag || (vnode.isComment && !vnode.asyncFactory)
}

export function addStyleToVnode(vnode, classes) {
	if (isIgnoredVnode(vnode)) {
		return vnode
	}

	if (!vnode.data) {
		vnode.data = {}
	}

	vnode.data.class = vnode.data.class ? toArray(vnode.data.class).concat(toArray(classes)) : toArray(classes)

	return vnode
}

import flatten from 'lodash/flatten'

export const colors = ['primary', 'info', 'success', 'warning', 'danger', 'dark', 'light']

export function style(node, ...styles) {
  if (isIgnoredVnode(node)) return node

  node.data = {
    ...node.data,
    class: flatten([styles, toArray(node.data.class)]),
  }

  return node
}
