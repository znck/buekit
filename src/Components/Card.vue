<script>
import PropTypes from '@znck/prop-types'
import { style, styleResolver, toArray } from '../utils'
import { createTag } from '../mixins'

export default {
  name: 'Card',
  functional: true,
  mixins: [createTag()],
  props: {
    title: PropTypes.string,
    centered: PropTypes.bool.value(false)
  },
  render (h, ctx) {
    const { centered, tag, title } = ctx.props
    const { default: contentSLOT, imageSLOT, title: headingSLOT, iconSLOT, linksSLOT } = ctx.slots()
    const _ = styleResolver(ctx.$style)

    const hasHeader = Array.isArray(headingSLOT) || title !== undefined || Array.isArray(iconSLOT)
    const headStyle = [_('card-header-title'), centered && _('is-centered')]
    const footLinkStyle = _('card-footer-item')

    return style(
      h(tag, ctx.data, [
        /* Render Header */
        hasHeader && h('header', { class: _('card-header') }, [
          headingSLOT ? style(headingSLOT[0], headStyle) : h('p', { class: headStyle }, title),
          iconSLOT && style(iconSLOT[0], _('card-header-icon'))
        ]),
        /* Render Image */
        imageSLOT && h('div', { class: _('card-image') }, imageSLOT),
        h('div', { class: _('card-content') }, contentSLOT),
        linksSLOT && h('footer', { class: _('card-footer') }, linksSLOT.map(link => style(link, footLinkStyle)))
      ]),
      _('card')
    )
  }
}
</script>

<style lang="scss" module>
@import '~bulma/sass/utilities/_all.sass';
@import '../_base.scss';
@import '~bulma/sass/components/card.sass';
</style>
