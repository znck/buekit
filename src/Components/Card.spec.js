import {shallow} from 'vue-test-utils'
import {createRenderer} from 'vue-server-renderer'

import Card from './Card.vue'

import { $style } from '../utils'

Card.$style = $style

describe('Card.vue', () => {
	it('matches snapshot', () => {
		const wrapper = shallow(Card, {})
		const renderer = createRenderer()

		renderer.renderToString(wrapper.vm, (err, str) => {
			if (err) {
				throw new Error(err)
			}
			expect(str).toMatchSnapshot()
		})
	})
})
