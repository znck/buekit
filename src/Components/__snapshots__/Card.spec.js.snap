// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card.vue matches snapshot 1`] = `
<div class="card">
    <div class="card-content"></div>
</div>
`;
